File: hive/server/hive_api/objects.py
"""Hive API: account, post, and comment object retrieval"""
import logging
from hive.server.hive_api.common import get_account_id, estimated_sp
log = logging.getLogger(__name__)

# Account objects
# ---------------

async def accounts_by_name(db, names, observer=None, lite=True):
    """Find and return accounts by `name`."""

    sql = """SELECT id, name, display_name, about, created_at,
                    vote_weight, rank, followers, following %s
               FROM hive_accounts WHERE name IN :names"""
    fields = '' if lite else ', location, website, profile_image, cover_image'
    rows = await db.query_all(sql % fields, names=tuple(names))

    accounts = {}
    for row in rows:
        account = {
            'id': row['id'],
            'name': row['name'],
            'created': str(row['created_at']).split(' ')[0],
            'sp': int(estimated_sp(row['vote_weight'])),
            'rank': row['rank'],
            'followers': row['followers'],
            'following': row['following'],
            'display_name': row['display_name'],
            'about': row['about'],
        }
        if not lite:
            account['location'] = row['location']
            account['website'] = row['website']
            account['profile_image'] = row['profile_image']
            account['cover_image'] = row['cover_image']
        accounts[account['id']] = account

    if observer:
        await _follow_contexts(db, accounts,
                               observer_id=await get_account_id(db, observer),
                               include_mute=not lite)

    return accounts.values()

async def _follow_contexts(db, accounts, observer_id, include_mute=False):
    sql = """SELECT following, state FROM hive_follows
              WHERE follower = :account_id AND following IN :ids"""
    rows = await db.query_all(sql,
                              account_id=observer_id,
                              ids=tuple(accounts.keys()))
    for row in rows:
        following_id = row[0]
        state = row[1]
        context = {'followed': state == 1}
        if include_mute and state == 2:
            context['muted'] = True
        accounts[following_id]['context'] = context

    for account in accounts.values():
        if 'context' not in account:
            account['context'] = {'followed': False}


# Comment objects
# ---------------

async def comments_by_id(db, ids, observer=None):
    """Given an array of post ids, returns comment objects keyed by id."""
    assert ids, 'no ids passed to comments_by_id'

    sql = """SELECT post_id, author, permlink, body, depth,
                    payout, payout_at, is_paidout, created_at, updated_at,
                    rshares, is_hidden, is_grayed, votes
               FROM hive_posts_cache WHERE post_id IN :ids""" #votes
    result = await db.query_all(sql, ids=tuple(ids))

    authors = set()
    by_id = {}
    for row in result:
        top_votes, observer_vote = _top_votes(row, 5, observer)
        post = {
            'id': row['post_id'],
            'author': row['author'],
            'url': row['author'] + '/' + row['permlink'],
            'depth': row['depth'],
            'body': row['body'],
            'payout': str(row['payout']),
            'created_at': str(row['created_at']),
            'updated_at': str(row['updated_at']),
            'payout_at': str(row['payout_at']),
            'is_paidout': row['is_paidout'],
            'rshares': row['rshares'],
            'hide': row['is_hidden'] or row['is_grayed'],
            'top_votes': top_votes,
        }

        authors.add(row['author'])

        if observer:
            post['context'] = {'vote_rshares': observer_vote}
        by_id[post['id']] = post

    by_id = await _append_flags(db, by_id)
    return {'posts': by_id, #[by_id[_id] for _id in ids],
            'accounts': await accounts_by_name(db, authors, observer, lite=True)}

async def posts_by_id(db, ids, observer=None, lite=True):
    """Given a list of post ids, returns lite post objects in the same order."""

    # pylint: disable=too-many-locals
    sql = """SELECT post_id, author, permlink, title, img_url, payout, promoted,
                    created_at, payout_at, is_nsfw, rshares, votes,
                    is_muted, is_invalid, %s
               FROM hive_posts_cache WHERE post_id IN :ids"""
    fields = ['preview'] if lite else ['body', 'updated_at', 'json']
    sql = sql % (', '.join(fields))

    reblogged_ids = await _reblogged_ids(db, observer, ids) if observer else []

    # TODO: filter out observer's mutes?

    # key by id.. returns sorted by input order
    authors = set()
    by_id = {}
    for row in await db.query_all(sql, ids=tuple(ids)):
        assert not row['is_muted']
        assert not row['is_invalid']
        pid = row['post_id']
        top_votes, observer_vote = _top_votes(row, 5, observer)

        obj = {
            'id': pid,
            'author': row['author'],
            'url': row['author'] + '/' + row['permlink'],
            'title': row['title'],
            'payout': float(row['payout']),
            'promoted': float(row['promoted']),
            'created_at': str(row['created_at']),
            'payout_at': str(row['payout_at']),
            'is_paidout': row['is_paidout'],
            'rshares' : row['rshares'],
            'top_votes' : top_votes,
            'thumb_url': row['img_url'],
            'is_nsfw' : row['is_nsfw'],
        }

        if lite:
            obj['preview'] = row['preview']
        else:
            obj['body'] = row['body']
            obj['updated_at'] = str(row['updated_at'])
            obj['json_metadata'] = row['json']

        if observer:
            obj['context'] = {
                'reblogged': obj['id'] in reblogged_ids,
                'vote_rshares': observer_vote
            }

        authors.add(obj['author'])
        by_id[row['post_id']] = obj

    # in rare cases of cache inconsistency, recover and warn
    missed = set(ids) - by_id.keys()
    if missed:
        log.warning("by_id do not exist in cache: %s", repr(missed))
        for _id in missed:
            ids.remove(_id)

    by_id = await _append_flags(db, by_id)
    return {'posts': [by_id[_id] for _id in ids],
            'accounts': await accounts_by_name(db, authors, observer, lite=True)}

async def _append_flags(db, posts):
    sql = """SELECT id, parent_id, community, category, is_muted, is_valid
               FROM hive_posts WHERE id IN :ids"""
    for row in await db.query_all(sql, ids=tuple(posts.keys())):
        post = posts[row['id']]
        post['parent_id'] = row['parent_id']
        post['community'] = row['community']
        post['category'] = row['category']
        post['is_muted'] = row['is_muted']
        post['is_valid'] = row['is_valid']
    return posts

async def _reblogged_ids(db, observer, post_ids):
    sql = """SELECT post_id FROM hive_reblogs
              WHERE account = :observer
                AND post_id IN :ids"""
    return  await db.query_col(sql, observer=observer, ids=tuple(post_ids))

def _top_votes(obj, limit, observer):
    observer_vote = None
    votes = []
    if obj['votes']:
        for csa in obj['votes'].split("\n"):
            print(">>>"+csa+"<<<<")
            voter, rshares = csa.split(",")[0:2]
            rshares = int(rshares)
            votes.append((voter, rshares))

            if observer == voter:
                observer_vote = rshares

    top = sorted(votes, key=lambda row: abs(int(row[1])), reverse=True)[:limit]

    return (top, observer_vote)

File: internal/api/hive/objects.go
package hive

import (
	"context"
	"database/sql"
	"sort"
	"strconv"
	"strings"
	"time"

	"gorm.io/gorm"

	"github.com/steemit/hivemind/internal/models"
)

// estimatedSp converts VESTS to SP units for display.
// Mirrors legacy hive_api/common.py estimated_sp.
func estimatedSp(vests float64) int64 {
	return int64(vests * 0.0005034)
}

// accountsByNames loads hive_api account objects by name.
// Mirrors legacy hive_api/objects.py accounts_by_name. When lite is false the
// profile fields (location, website, images) are included. An optional
// observer adds followed/muted context to each account.
func accountsByNames(ctx context.Context, db *gorm.DB, names []string, observer string, lite bool) ([]map[string]interface{}, error) {
	if len(names) == 0 {
		return []map[string]interface{}{}, nil
	}

	var accounts []models.Account
	if err := db.WithContext(ctx).
		Where("name IN ?", names).
		Find(&accounts).Error; err != nil {
		return nil, err
	}

	byID := make(map[int64]map[string]interface{}, len(accounts))
	for _, acc := range accounts {
		obj := map[string]interface{}{
			"id":           acc.ID,
			"name":         acc.Name,
			"created":      acc.CreatedAt.Format("2006-01-02"),
			"sp":           estimatedSp(acc.VoteWeight),
			"rank":         acc.Rank,
			"followers":    acc.Followers,
			"following":    acc.Following,
			"display_name": nullString(acc.DisplayName),
			"about":        nullString(acc.About),
		}
		if !lite {
			obj["location"] = nullString(acc.Location)
			obj["website"] = nullString(acc.Website)
			obj["profile_image"] = acc.ProfileImage
			obj["cover_image"] = acc.CoverImage
		}
		byID[acc.ID] = obj
	}

	if observer != "" {
		if err := appendFollowContexts(ctx, db, byID, observer, !lite); err != nil {
			return nil, err
		}
	}

	// Preserve the input names order (legacy returns rows in query order;
	// hive_api callers rely on list semantics).
	out := make([]map[string]interface{}, 0, len(names))
	for _, name := range names {
		for _, acc := range accounts {
			if acc.Name == name {
				out = append(out, byID[acc.ID])
				break
			}
		}
	}
	return out, nil
}

// appendFollowContexts adds {followed, muted} context for each account
// relative to `observer`. Mirrors legacy _follow_contexts.
func appendFollowContexts(ctx context.Context, db *gorm.DB, accounts map[int64]map[string]interface{}, observer string, includeMute bool) error {
	var observerAcc models.Account
	if err := db.WithContext(ctx).
		Where("name = ?", observer).
		Select("id").
		First(&observerAcc).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil
		}
		return err
	}

	ids := make([]int64, 0, len(accounts))
	for id := range accounts {
		ids = append(ids, id)
	}
	if len(ids) == 0 {
		return nil
	}

	var follows []models.Follow
	if err := db.WithContext(ctx).
		Where("follower = ? AND following IN ?", observerAcc.ID, ids).
		Find(&follows).Error; err != nil {
		return err
	}

	for _, f := range follows {
		acc, ok := accounts[f.FollowingID]
		if !ok {
			continue
		}
		contextObj := map[string]interface{}{"followed": f.State == 1 || f.State == 3}
		if includeMute && (f.State == 2 || f.State == 3) {
			contextObj["muted"] = true
		}
		acc["context"] = contextObj
	}
	for _, acc := range accounts {
		if _, ok := acc["context"]; !ok {
			acc["context"] = map[string]interface{}{"followed": false}
		}
	}
	return nil
}

// postsByID loads lite hive_api post objects by ID, returned in input order
// together with the distinct authors. Mirrors legacy posts_by_id
// (hive_api/objects.py): {"posts": [...], "accounts": [...]}.
func postsByID(ctx context.Context, db *gorm.DB, ids []int64, observer string, lite bool) (map[string]interface{}, error) {
	if len(ids) == 0 {
		return map[string]interface{}{
			"posts":    []map[string]interface{}{},
			"accounts": []map[string]interface{}{},
		}, nil
	}

	var caches []models.PostCache
	if err := db.WithContext(ctx).
		Where("post_id IN ?", ids).
		Find(&caches).Error; err != nil {
		return nil, err
	}
	cacheMap := make(map[int64]*models.PostCache, len(caches))
	for i := range caches {
		cacheMap[caches[i].PostID] = &caches[i]
	}

	var posts []models.Post
	if err := db.WithContext(ctx).
		Where("id IN ?", ids).
		Find(&posts).Error; err != nil {
		return nil, err
	}
	postMap := make(map[int64]*models.Post, len(posts))
	for i := range posts {
		postMap[posts[i].ID] = &posts[i]
	}

	authorSet := make(map[string]bool)
	byID := make(map[int64]map[string]interface{}, len(ids))
	for _, id := range ids {
		post, cache := postMap[id], cacheMap[id]
		if post == nil || cache == nil {
			continue
		}

		obj := map[string]interface{}{
			"id":         id,
			"author":     cache.Author,
			"url":        cache.Author + "/" + cache.Permlink,
			"title":      cache.Title,
			"payout":     cache.Payout,
			"promoted":   cache.Promoted,
			"created_at": cache.CreatedAt.Format(time.RFC3339),
			"payout_at":  cache.PayoutAt.Format(time.RFC3339),
			"is_paidout": cache.IsPaidout,
			"rshares":    cache.RShares,
			"top_votes":  topVotes(cache.Votes, 5),
			"thumb_url":  cache.ImgURL,
			"is_nsfw":    cache.IsNSFW,
		}
		if lite {
			obj["preview"] = cache.Preview
		} else {
			obj["body"] = cache.Body
			obj["updated_at"] = cache.UpdatedAt.Format(time.RFC3339)
			obj["json_metadata"] = cache.JSON
		}
		obj["parent_id"] = nullInt64(post.ParentID)
		obj["community_id"] = nullInt64(post.CommunityID)
		obj["category"] = post.Category
		obj["is_muted"] = post.IsMuted
		obj["is_valid"] = post.IsValid

		authorSet[cache.Author] = true
		byID[id] = obj
	}

	// Recover from cache inconsistency by dropping missing ids (legacy warns).
	orderedIDs := make([]int64, 0, len(ids))
	for _, id := range ids {
		if _, ok := byID[id]; ok {
			orderedIDs = append(orderedIDs, id)
		}
	}

	postList := make([]map[string]interface{}, 0, len(orderedIDs))
	for _, id := range orderedIDs {
		postList = append(postList, byID[id])
	}

	authors := make([]string, 0, len(authorSet))
	for name := range authorSet {
		authors = append(authors, name)
	}
	sort.Strings(authors)

	accounts, err := accountsByNames(ctx, db, authors, observer, true)
	if err != nil {
		return nil, err
	}

	return map[string]interface{}{
		"posts":    postList,
		"accounts": accounts,
	}, nil
}

// topVotes parses the votes CSV and returns the `limit` largest by abs
// rshares as [voter, rshares] pairs. Mirrors legacy _top_votes.
func topVotes(voteCSV string, limit int) []interface{} {
	if voteCSV == "" {
		return []interface{}{}
	}
	type vote struct {
		voter   string
		rshares int64
	}
	votes := make([]vote, 0, 16)
	for _, line := range strings.Split(voteCSV, "\n") {
		parts := strings.Split(line, ",")
		if len(parts) < 2 {
			continue
		}
		rshares, err := strconv.ParseInt(parts[1], 10, 64)
		if err != nil {
			continue
		}
		votes = append(votes, vote{voter: parts[0], rshares: rshares})
	}
	sort.SliceStable(votes, func(i, j int) bool {
		return abs64(votes[i].rshares) > abs64(votes[j].rshares)
	})
	if len(votes) > limit {
		votes = votes[:limit]
	}
	out := make([]interface{}, 0, len(votes))
	for _, v := range votes {
		out = append(out, []interface{}{v.voter, v.rshares})
	}
	return out
}

func abs64(v int64) int64 {
	if v < 0 {
		return -v
	}
	return v
}

// nullString unwraps sql.NullString for JSON output (nil when NULL).
func nullString(ns sql.NullString) interface{} {
	if ns.Valid {
		return ns.String
	}
	return nil
}

// nullInt64 unwraps sql.NullInt64 for JSON output (nil when NULL).
func nullInt64(v sql.NullInt64) interface{} {
	if v.Valid {
		return v.Int64
	}
	return nil
}
